File: swim-system-java/swim-core-java/swim.recon/src/main/java/swim/recon/ReconStructureWriter.java
// Copyright 2015-2021 Swim inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package swim.recon;

import java.util.Iterator;
import swim.codec.Output;
import swim.codec.Writer;
import swim.codec.WriterException;
import swim.structure.Absent;
import swim.structure.Attr;
import swim.structure.Bool;
import swim.structure.Data;
import swim.structure.Expression;
import swim.structure.Extant;
import swim.structure.Field;
import swim.structure.Func;
import swim.structure.Item;
import swim.structure.Num;
import swim.structure.Operator;
import swim.structure.Record;
import swim.structure.Selector;
import swim.structure.Slot;
import swim.structure.Text;
import swim.structure.Value;
import swim.structure.func.BridgeFunc;
import swim.structure.func.LambdaFunc;
import swim.structure.operator.BinaryOperator;
import swim.structure.operator.ConditionalOperator;
import swim.structure.operator.InvokeOperator;
import swim.structure.operator.UnaryOperator;
import swim.structure.selector.ChildrenSelector;
import swim.structure.selector.DescendantsSelector;
import swim.structure.selector.FilterSelector;
import swim.structure.selector.GetAttrSelector;
import swim.structure.selector.GetItemSelector;
import swim.structure.selector.GetSelector;
import swim.structure.selector.IdentitySelector;
import swim.structure.selector.KeysSelector;
import swim.structure.selector.LiteralSelector;
import swim.structure.selector.ValuesSelector;

public class ReconStructureWriter extends ReconWriter<Item, Value> {

  @Override
  public boolean isField(Item item) {
    return item instanceof Field;
  }

  @Override
  public boolean isAttr(Item item) {
    return item instanceof Attr;
  }

  @Override
  public boolean isSlot(Item item) {
    return item instanceof Slot;
  }

  @Override
  public boolean isValue(Item item) {
    return item instanceof Value;
  }

  @Override
  public boolean isRecord(Item item) {
    return item instanceof Record;
  }

  @Override
  public boolean isText(Item item) {
    return item instanceof Text;
  }

  @Override
  public boolean isNum(Item item) {
    return item instanceof Num;
  }

  @Override
  public boolean isBool(Item item) {
    return item instanceof Bool;
  }

  @Override
  public boolean isExpression(Item item) {
    return item instanceof Expression;
  }

  @Override
  public boolean isExtant(Item item) {
    return item instanceof Extant;
  }

  @Override
  public Iterator<Item> items(Item item) {
    return item.iterator();
  }

  @Override
  public Item item(Value value) {
    return value;
  }

  @Override
  public Value key(Item item) {
    return item.key();
  }

  @Override
  public Value value(Item item) {
    return item.toValue();
  }

  @Override
  public String string(Item item) {
    return item.stringValue();
  }

  @Override
  public int precedence(Item item) {
    return item.precedence();
  }

  @Override
  public int sizeOfItem(Item item) {
    if (item instanceof Field) {
      if (item instanceof Attr) {
        final Attr that = (Attr) item;
        return this.sizeOfAttr(that.key(), that.value());
      } else if (item instanceof Slot) {
        final Slot that = (Slot) item;
        return this.sizeOfSlot(that.key(), that.value());
      }
    } else if (item instanceof Value) {
      return this.sizeOfValue((Value) item);
    }
    throw new WriterException("No Recon serialization for " + item);
  }

  @Override
  public Writer<?, ?> writeItem(Item item, Output<?> output) {
    if (item instanceof Field) {
      if (item instanceof Attr) {
        final Attr that = (Attr) item;
        return this.writeAttr(that.key(), that.value(), output);
      } else if (item instanceof Slot) {
        final Slot that = (Slot) item;
        return this.writeSlot(that.key(), that.value(), output);
      }
    } else if (item instanceof Value) {
      return this.writeValue((Value) item, output);
    }
    return Writer.error(new WriterException("No Recon serialization for " + item));
  }

  @Override
  public int sizeOfValue(Value value) {
    if (value instanceof Record) {
      final Record that = (Record) value;
      return this.sizeOfRecord(that);
    } else if (value instanceof Data) {
      final Data that = (Data) value;
      return this.sizeOfData(that.size());
    } else if (value instanceof Text) {
      final Text that = (Text) value;
      return this.sizeOfText(that.stringValue());
    } else if (value instanceof Num) {
      final Num that = (Num) value;
      if (that.isUint32()) {
        return this.sizeOfUint32(that.intValue());
      } else if (that.isUint64()) {
        return this.sizeOfUint64(that.longValue());
      } else if (that.isValidInt()) {
        return this.sizeOfNum(that.intValue());
      } else if (that.isValidLong()) {
        return this.sizeOfNum(that.longValue());
      } else if (that.isValidFloat()) {
        return this.sizeOfNum(that.floatValue());
      } else if (that.isValidDouble()) {
        return this.sizeOfNum(that.doubleValue());
      } else if (that.isValidInteger()) {
        return this.sizeOfNum(that.integerValue());
      }
    } else if (value instanceof Bool) {
      final Bool that = (Bool) value;
      return this.sizeOfBool(that.booleanValue());
    } else if (value instanceof Selector) {
      return this.sizeOfSelector((Selector) value);
    } else if (value instanceof Operator) {
      return this.sizeOfOperator((Operator) value);
    } else if (value instanceof Func) {
      return this.sizeOfFunc((Func) value);
    } else if (value instanceof Extant) {
      return this.sizeOfExtant();
    } else if (value instanceof Absent) {
      return this.sizeOfAbsent();
    }
    throw new WriterException("No Recon serialization for " + value);
  }

  @Override
  public Writer<?, ?> writeValue(Value value, Output<?> output) {
    if (value instanceof Record) {
      final Record that = (Record) value;
      return this.writeRecord(that, output);
    } else if (value instanceof Data) {
      final Data that = (Data) value;
      return this.writeData(that.asByteBuffer(), output);
    } else if (value instanceof Text) {
      final Text that = (Text) value;
      return this.writeText(that.stringValue(), output);
    } else if (value instanceof Num) {
      final Num that = (Num) value;
      if (that.isUint32()) {
        return this.writeUint32(that.intValue(), output);
      } else if (that.isUint64()) {
        return this.writeUint64(that.longValue(), output);
      } else if (that.isValidInt()) {
        return this.writeNum(that.intValue(), output);
      } else if (that.isValidLong()) {
        return this.writeNum(that.longValue(), output);
      } else if (that.isValidFloat()) {
        return this.writeNum(that.floatValue(), output);
      } else if (that.isValidDouble()) {
        return this.writeNum(that.doubleValue(), output);
      } else if (that.isValidInteger()) {
        return this.writeNum(that.integerValue(), output);
      }
    } else if (value instanceof Bool) {
      final Bool that = (Bool) value;
      return this.writeBool(that.booleanValue(), output);
    } else if (value instanceof Selector) {
      return this.writeSelector((Selector) value, output);
    } else if (value instanceof Operator) {
      return this.writeOperator((Operator) value, output);
    } else if (value instanceof Func) {
      return this.writeFunc((Func) value, output);
    } else if (value instanceof Extant) {
      return this.writeExtant(output);
    } else if (value instanceof Absent) {
      return this.writeAbsent(output);
    }
    return Writer.error(new WriterException("No Recon serialization for " + value));
  }

  public int sizeOfSelector(Selector selector) {
    if (selector instanceof IdentitySelector) {
      return this.sizeOfIdentitySelector();
    } else if (selector instanceof LiteralSelector) {
      final LiteralSelector that = (LiteralSelector) selector;
      return this.sizeOfLiteralSelector(that.item(), that.then());
    } else if (selector instanceof GetSelector) {
      final GetSelector that = (GetSelector) selector;
      return this.sizeOfGetSelector(that.accessor(), that.then());
    } else if (selector instanceof GetAttrSelector) {
      final GetAttrSelector that = (GetAttrSelector) selector;
      return this.sizeOfGetAttrSelector(that.accessor(), that.then());
    } else if (selector instanceof GetItemSelector) {
      final GetItemSelector that = (GetItemSelector) selector;
      return this.sizeOfGetItemSelector(that.accessor(), that.then());
    } else if (selector instanceof KeysSelector) {
      final KeysSelector that = (KeysSelector) selector;
      return this.sizeOfKeysSelector(that.then());
    } else if (selector instanceof ValuesSelector) {
      final ValuesSelector that = (ValuesSelector) selector;
      return this.sizeOfValuesSelector(that.then());
    } else if (selector instanceof ChildrenSelector) {
      final ChildrenSelector that = (ChildrenSelector) selector;
      return this.sizeOfChildrenSelector(that.then());
    } else if (selector instanceof DescendantsSelector) {
      final DescendantsSelector that = (DescendantsSelector) selector;
      return this.sizeOfDescendantsSelector(that.then());
    } else if (selector instanceof FilterSelector) {
      final FilterSelector that = (FilterSelector) selector;
      return this.sizeOfFilterSelector(that.predicate(), that.then());
    }
    throw new WriterException("No Recon serialization for " + selector);
  }

  public Writer<?, ?> writeSelector(Selector selector, Output<?> output) {
    if (selector instanceof IdentitySelector) {
      return this.writeIdentitySelector(output);
    } else if (selector instanceof LiteralSelector) {
      final LiteralSelector that = (LiteralSelector) selector;
      return this.writeLiteralSelector(that.item(), that.then(), output);
    } else if (selector instanceof GetSelector) {
      final GetSelector that = (GetSelector) selector;
      return this.writeGetSelector(that.accessor(), that.then(), output);
    } else if (selector instanceof GetAttrSelector) {
      final GetAttrSelector that = (GetAttrSelector) selector;
      return this.writeGetAttrSelector(that.accessor(), that.then(), output);
    } else if (selector instanceof GetItemSelector) {
      final GetItemSelector that = (GetItemSelector) selector;
      return this.writeGetItemSelector(that.accessor(), that.then(), output);
    } else if (selector instanceof KeysSelector) {
      final KeysSelector that = (KeysSelector) selector;
      return this.writeKeysSelector(that.then(), output);
    } else if (selector instanceof ValuesSelector) {
      final ValuesSelector that = (ValuesSelector) selector;
      return this.writeValuesSelector(that.then(), output);
    } else if (selector instanceof ChildrenSelector) {
      final ChildrenSelector that = (ChildrenSelector) selector;
      return this.writeChildrenSelector(that.then(), output);
    } else if (selector instanceof DescendantsSelector) {
      final DescendantsSelector that = (DescendantsSelector) selector;
      return this.writeDescendantsSelector(that.then(), output);
    } else if (selector instanceof FilterSelector) {
      final FilterSelector that = (FilterSelector) selector;
      return this.writeFilterSelector(that.predicate(), that.then(), output);
    }
    return Writer.error(new WriterException("No Recon serialization for " + selector));
  }

  public int sizeOfOperator(Operator operator) {
    if (operator instanceof BinaryOperator) {
      final BinaryOperator that = (BinaryOperator) operator;
      return this.sizeOfInfixOperator(that.lhs(), that.operator(), that.rhs(), that.precedence());
    } else if (operator instanceof UnaryOperator) {
      final UnaryOperator that = (UnaryOperator) operator;
      return this.sizeOfPrefixOperator(that.operator(), that.operand(), that.precedence());
    } else if (operator instanceof InvokeOperator) {
      final InvokeOperator that = (InvokeOperator) operator;
      return this.sizeOfInvokeOperator(that.func(), that.args());
    } else if (operator instanceof ConditionalOperator) {
      final ConditionalOperator that = (ConditionalOperator) operator;
      return this.sizeOfConditionalOperator(that.ifTerm(), that.thenTerm(), that.elseTerm(), that.precedence());
    }
    throw new WriterException("No Recon serialization for " + operator);
  }

  public Writer<?, ?> writeOperator(Operator operator, Output<?> output) {
    if (operator instanceof BinaryOperator) {
      final BinaryOperator that = (BinaryOperator) operator;
      return this.writeInfixOperator(that.lhs(), that.operator(), that.rhs(), that.precedence(), output);
    } else if (operator instanceof UnaryOperator) {
      final UnaryOperator that = (UnaryOperator) operator;
      return this.writePrefixOperator(that.operator(), that.operand(), that.precedence(), output);
    } else if (operator instanceof InvokeOperator) {
      final InvokeOperator that = (InvokeOperator) operator;
      return this.writeInvokeOperator(that.func(), that.args(), output);
    } else if (operator instanceof ConditionalOperator) {
      final ConditionalOperator that = (ConditionalOperator) operator;
      return this.writeConditionalOperator(that.ifTerm(), that.thenTerm(), that.elseTerm(), that.precedence(), output);
    }
    return Writer.error(new WriterException("No Recon serialization for " + operator));
  }

  public int sizeOfFunc(Func func) {
    if (func instanceof LambdaFunc) {
      final LambdaFunc that = (LambdaFunc) func;
      return this.sizeOfLambdaFunc(that.bindings(), that.template());
    } else if (func instanceof BridgeFunc) {
      return 0;
    }
    throw new WriterException("No Recon serialization for " + func);
  }

  public Writer<?, ?> writeFunc(Func func, Output<?> output) {
    if (func instanceof LambdaFunc) {
      final LambdaFunc that = (LambdaFunc) func;
      return this.writeLambdaFunc(that.bindings(), that.template(), output);
    } else if (func instanceof BridgeFunc) {
      return Writer.done();
    }
    return Writer.error(new WriterException("No Recon serialization for " + func));
  }

  @Override
  public int sizeOfBlockItem(Item item) {
    if (item instanceof Field) {
      return this.sizeOfItem(item);
    } else if (item instanceof Value) {
      return this.sizeOfBlockValue((Value) item);
    }
    throw new WriterException("No Recon serialization for " + item);
  }

  @Override
  public Writer<?, ?> writeBlockItem(Item item, Output<?> output) {
    if (item instanceof Field) {
      return this.writeItem(item, output);
    } else if (item instanceof Value) {
      return this.writeBlockValue((Value) item, output);
    }
    return Writer.error(new WriterException("No Recon serialization for " + item));
  }

  @Override
  public int sizeOfBlockValue(Value value) {
    if (value instanceof Record) {
      return this.sizeOfBlock((Record) value);
    }
    return this.sizeOfValue(value);
  }

  @Override
  public Writer<?, ?> writeBlockValue(Value value, Output<?> output) {
    if (value instanceof Record) {
      return this.writeBlock((Record) value, output);
    }
    return this.writeValue(value, output);
  }

  @Override
  public int sizeOfThen(Value then) {
    if (then instanceof Selector) {
      final Selector selector = (Selector) then;
      if (selector instanceof IdentitySelector) {
        return this.sizeOfThenIdentitySelector();
      } else if (selector instanceof LiteralSelector) {
        final LiteralSelector that = (LiteralSelector) selector;
        return this.sizeOfThenLiteralSelector(that.item(), that.then());
      } else if (selector instanceof GetSelector) {
        final GetSelector that = (GetSelector) selector;
        return this.sizeOfThenGetSelector(that.accessor(), that.then());
      } else if (selector instanceof GetAttrSelector) {
        final GetAttrSelector that = (GetAttrSelector) selector;
        return this.sizeOfThenGetAttrSelector(that.accessor(), that.then());
      } else if (selector instanceof GetItemSelector) {
        final GetItemSelector that = (GetItemSelector) selector;
        return this.sizeOfThenGetItemSelector(that.accessor(), that.then());
      } else if (selector instanceof KeysSelector) {
        final KeysSelector that = (KeysSelector) selector;
        return this.sizeOfThenKeysSelector(that.then());
      } else if (selector instanceof ValuesSelector) {
        final ValuesSelector that = (ValuesSelector) selector;
        return this.sizeOfThenValuesSelector(that.then());
      } else if (selector instanceof ChildrenSelector) {
        final ChildrenSelector that = (ChildrenSelector) selector;
        return this.sizeOfThenChildrenSelector(that.then());
      } else if (selector instanceof DescendantsSelector) {
        final DescendantsSelector that = (DescendantsSelector) selector;
        return this.sizeOfThenDescendantsSelector(that.then());
      } else if (selector instanceof FilterSelector) {
        final FilterSelector that = (FilterSelector) selector;
        return this.sizeOfThenFilterSelector(that.predicate(), that.then());
      }
    }
    throw new WriterException("No Recon serialization for " + then);
  }

  @Override
  public Writer<?, ?> writeThen(Value then, Output<?> output) {
    if (then instanceof Selector) {
      final Selector selector = (Selector) then;
      if (selector instanceof IdentitySelector) {
        return this.writeThenIdentitySelector(output);
      } else if (selector instanceof LiteralSelector) {
        final LiteralSelector that = (LiteralSelector) selector;
        return this.writeThenLiteralSelector(that.item(), that.then(), output);
      } else if (selector instanceof GetSelector) {
        final GetSelector that = (GetSelector) selector;
        return this.writeThenGetSelector(that.accessor(), that.then(), output);
      } else if (selector instanceof GetAttrSelector) {
        final GetAttrSelector that = (GetAttrSelector) selector;
        return this.writeThenGetAttrSelector(that.accessor(), that.then(), output);
      } else if (selector instanceof GetItemSelector) {
        final GetItemSelector that = (GetItemSelector) selector;
        return this.writeThenGetItemSelector(that.accessor(), that.then(), output);
      } else if (selector instanceof KeysSelector) {
        final KeysSelector that = (KeysSelector) selector;
        return this.writeThenKeysSelector(that.then(), output);
      } else if (selector instanceof ValuesSelector) {
        final ValuesSelector that = (ValuesSelector) selector;
        return this.writeThenValuesSelector(that.then(), output);
      } else if (selector instanceof ChildrenSelector) {
        final ChildrenSelector that = (ChildrenSelector) selector;
        return this.writeThenChildrenSelector(that.then(), output);
      } else if (selector instanceof DescendantsSelector) {
        final DescendantsSelector that = (DescendantsSelector) selector;
        return this.writeThenDescendantsSelector(that.then(), output);
      } else if (selector instanceof FilterSelector) {
        final FilterSelector that = (FilterSelector) selector;
        return this.writeThenFilterSelector(that.predicate(), that.then(), output);
      }
    }
    return Writer.error(new WriterException("No Recon serialization for " + then));
  }

}

File: swim-js/swim-runtime/swim-core/swim-recon/src/main/writer/ReconStructureWriter.ts
// Copyright 2015-2022 Swim.inc
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import type {Cursor} from "@swim/util";
import {Output, WriterException, Writer} from "@swim/codec";
import {
  Item,
  Field,
  Attr,
  Slot,
  Value,
  Record,
  Data,
  Text,
  Num,
  Bool,
  Expression,
  Operator,
  Selector,
  Func,
  Extant,
  Absent,
  ChildrenSelector,
  DescendantsSelector,
  FilterSelector,
  GetAttrSelector,
  GetItemSelector,
  GetSelector,
  IdentitySelector,
  KeysSelector,
  LiteralSelector,
  ValuesSelector,
  BinaryOperator,
  ConditionalOperator,
  InvokeOperator,
  UnaryOperator,
  BridgeFunc,
  LambdaFunc,
} from "@swim/structure";
import {ReconWriter} from "./ReconWriter";

/** @public */
export class ReconStructureWriter extends ReconWriter<Item, Value> {
  override isField(item: Item): boolean {
    return item instanceof Field;
  }

  override isAttr(item: Item): boolean {
    return item instanceof Attr;
  }

  override isSlot(item: Item): boolean {
    return item instanceof Slot;
  }

  override isValue(item: Item): boolean {
    return item instanceof Value;
  }

  override isRecord(item: Item): boolean {
    return item instanceof Record;
  }

  override isText(item: Item): boolean {
    return item instanceof Text;
  }

  override isNum(item: Item): boolean {
    return item instanceof Num;
  }

  override isBool(item: Item): boolean {
    return item instanceof Bool;
  }

  override isExpression(item: Item): boolean {
    return item instanceof Expression;
  }

  override isExtant(item: Item): boolean {
    return item instanceof Extant;
  }

  override items(item: Item): Cursor<Item> {
    return item.iterator();
  }

  override item(value: Value): Item {
    return value;
  }

  override key(item: Item): Value {
    return item.key;
  }

  override value(item: Item): Value {
    return item.toValue();
  }

  override string(item: Item): string {
    return item.stringValue("");
  }

  override precedence(item: Item): number {
    return item.precedence;
  }

  override sizeOfItem(item: Item): number {
    if (item instanceof Field) {
      if (item instanceof Attr) {
        return this.sizeOfAttr(item.key, item.value);
      } else if (item instanceof Slot) {
        return this.sizeOfSlot(item.key, item.value);
      }
    } else if (item instanceof Value) {
      return this.sizeOfValue(item);
    }
    throw new WriterException("No Recon serialization for " + item);
  }

  override writeItem(output: Output, item: Item): Writer {
    if (item instanceof Field) {
      if (item instanceof Attr) {
        return this.writeAttr(output, item.key, item.value);
      } else if (item instanceof Slot) {
        return this.writeSlot(output, item.key, item.value);
      }
    } else if (item instanceof Value) {
      return this.writeValue(output, item);
    }
    return Writer.error(new WriterException("No Recon serialization for " + item));
  }

  override sizeOfValue(value: Value): number {
    if (value instanceof Record) {
      return this.sizeOfRecord(value);
    } else if (value instanceof Data) {
      return this.sizeOfData(value.size);
    } else if (value instanceof Text) {
      return this.sizeOfText(value.value);
    } else if (value instanceof Num) {
      if (value.isUint32()) {
        return this.sizeOfUint32(value.value);
      } else if (value.isUint64()) {
        return this.sizeOfUint64(value.value);
      } else {
        return this.sizeOfNum(value.value);
      }
    } else if (value instanceof Bool) {
      return this.sizeOfBool(value.value);
    } else if (value instanceof Selector) {
      return this.sizeOfSelector(value);
    } else if (value instanceof Operator) {
      return this.sizeOfOperator(value);
    } else if (value instanceof Func) {
      return this.sizeOfFunc(value);
    } else if (value instanceof Extant) {
      return this.sizeOfExtant();
    } else if (value instanceof Absent) {
      return this.sizeOfAbsent();
    }
    throw new WriterException("No Recon serialization for " + value);
  }

  override writeValue(output: Output, value: Value): Writer {
    if (value instanceof Record) {
      return this.writeRecord(output, value);
    } else if (value instanceof Data) {
      return this.writeData(output, value.asUint8Array());
    } else if (value instanceof Text) {
      return this.writeText(output, value.value);
    } else if (value instanceof Num) {
      if (value.isUint32()) {
        return this.writeUint32(output, value.value);
      } else if (value.isUint64()) {
        return this.writeUint64(output, value.value);
      } else {
        return this.writeNum(output, value.value);
      }
    } else if (value instanceof Bool) {
      return this.writeBool(output, value.value);
    } else if (value instanceof Selector) {
      return this.writeSelector(output, value);
    } else if (value instanceof Operator) {
      return this.writeOperator(output, value);
    } else if (value instanceof Func) {
      return this.writeFunc(output, value);
    } else if (value instanceof Extant) {
      return this.writeExtant(output);
    } else if (value instanceof Absent) {
      return this.writeAbsent(output);
    }
    return Writer.error(new WriterException("No Recon serialization for " + value));
  }

  sizeOfSelector(selector: Selector): number {
    if (selector instanceof IdentitySelector) {
      return this.sizeOfIdentitySelector();
    } else if (selector instanceof LiteralSelector) {
      return this.sizeOfLiteralSelector(selector.item, selector.then);
    } else if (selector instanceof GetSelector) {
      return this.sizeOfGetSelector(selector.item, selector.then);
    } else if (selector instanceof GetAttrSelector) {
      return this.sizeOfGetAttrSelector(selector.item, selector.then);
    } else if (selector instanceof GetItemSelector) {
      return this.sizeOfGetItemSelector(selector.item, selector.then);
    } else if (selector instanceof KeysSelector) {
      return this.sizeOfKeysSelector(selector.then);
    } else if (selector instanceof ValuesSelector) {
      return this.sizeOfValuesSelector(selector.then);
    } else if (selector instanceof ChildrenSelector) {
      return this.sizeOfChildrenSelector(selector.then);
    } else if (selector instanceof DescendantsSelector) {
      return this.sizeOfDescendantsSelector(selector.then);
    } else if (selector instanceof FilterSelector) {
      return this.sizeOfFilterSelector(selector.predicate, selector.then);
    }
    throw new WriterException("No Recon serialization for " + selector);
  }

  writeSelector(output: Output, selector: Selector): Writer {
    if (selector instanceof IdentitySelector) {
      return this.writeIdentitySelector(output);
    } else if (selector instanceof LiteralSelector) {
      return this.writeLiteralSelector(output, selector.item, selector.then);
    } else if (selector instanceof GetSelector) {
      return this.writeGetSelector(output, selector.item, selector.then);
    } else if (selector instanceof GetAttrSelector) {
      return this.writeGetAttrSelector(output, selector.item, selector.then);
    } else if (selector instanceof GetItemSelector) {
      return this.writeGetItemSelector(output, selector.item, selector.then);
    } else if (selector instanceof KeysSelector) {
      return this.writeKeysSelector(output, selector.then);
    } else if (selector instanceof ValuesSelector) {
      return this.writeValuesSelector(output, selector.then);
    } else if (selector instanceof ChildrenSelector) {
      return this.writeChildrenSelector(output, selector.then);
    } else if (selector instanceof DescendantsSelector) {
      return this.writeDescendantsSelector(output, selector.then);
    } else if (selector instanceof FilterSelector) {
      return this.writeFilterSelector(output, selector.predicate, selector.then);
    }
    return Writer.error(new WriterException("No Recon serialization for " + selector));
  }

  sizeOfOperator(operator: Operator): number {
    if (operator instanceof BinaryOperator) {
      return this.sizeOfInfixOperator(operator.operand1, operator.operator, operator.operand2, operator.precedence);
    } else if (operator instanceof UnaryOperator) {
      return this.sizeOfPrefixOperator(operator.operator, operator.operand, operator.precedence);
    } else if (operator instanceof InvokeOperator) {
      return this.sizeOfInvokeOperator(operator.func, operator.args);
    } else if (operator instanceof ConditionalOperator) {
      return this.sizeOfConditionalOperator(operator.ifTerm, operator.thenTerm, operator.elseTerm, operator.precedence);
    }
    throw new WriterException("No Recon serialization for " + operator);
  }

  writeOperator(output: Output, operator: Operator): Writer {
    if (operator instanceof BinaryOperator) {
      return this.writeInfixOperator(output, operator.operand1, operator.operator, operator.operand2, operator.precedence);
    } else if (operator instanceof UnaryOperator) {
      return this.writePrefixOperator(output, operator.operator, operator.operand, operator.precedence);
    } else if (operator instanceof InvokeOperator) {
      return this.writeInvokeOperator(output, operator.func, operator.args);
    } else if (operator instanceof ConditionalOperator) {
      return this.writeConditionalOperator(output, operator.ifTerm, operator.thenTerm, operator.elseTerm, operator.precedence);
    }
    return Writer.error(new WriterException("No Recon serialization for " + operator));
  }

  sizeOfFunc(func: Func): number {
    if (func instanceof LambdaFunc) {
      return this.sizeOfLambdaFunc(func.bindings, func.template);
    } else if (func instanceof BridgeFunc) {
      return 0;
    }
    throw new WriterException("No Recon serialization for " + func);
  }

  writeFunc(output: Output, func: Func): Writer {
    if (func instanceof LambdaFunc) {
      return this.writeLambdaFunc(output, func.bindings, func.template);
    } else if (func instanceof BridgeFunc) {
      return Writer.end();
    }
    return Writer.error(new WriterException("No Recon serialization for " + func));
  }

  override sizeOfBlockItem(item: Item): number {
    if (item instanceof Field) {
      return this.sizeOfItem(item);
    } else if (item instanceof Value) {
      return this.sizeOfBlockValue(item);
    }
    throw new WriterException("No Recon serialization for " + item);
  }

  override writeBlockItem(output: Output, item: Item): Writer {
    if (item instanceof Field) {
      return this.writeItem(output, item);
    } else if (item instanceof Value) {
      return this.writeBlockValue(output, item);
    }
    return Writer.error(new WriterException("No Recon serialization for " + item));
  }

  override sizeOfBlockValue(value: Value): number {
    if (value instanceof Record) {
      return this.sizeOfBlock(value);
    }
    return this.sizeOfValue(value);
  }

  override writeBlockValue(output: Output, value: Value): Writer {
    if (value instanceof Record) {
      return this.writeBlock(output, value);
    }
    return this.writeValue(output, value);
  }

  override sizeOfThen(then: Value): number {
    if (then instanceof Selector) {
      if (then instanceof IdentitySelector) {
        return this.sizeOfThenIdentitySelector();
      } else if (then instanceof LiteralSelector) {
        return this.sizeOfThenLiteralSelector(then.item, then.then);
      } else if (then instanceof GetSelector) {
        return this.sizeOfThenGetSelector(then.item, then.then);
      } else if (then instanceof GetAttrSelector) {
        return this.sizeOfThenGetAttrSelector(then.item, then.then);
      } else if (then instanceof GetItemSelector) {
        return this.sizeOfThenGetItemSelector(then.item, then.then);
      } else if (then instanceof KeysSelector) {
        return this.sizeOfThenKeysSelector(then.then);
      } else if (then instanceof ValuesSelector) {
        return this.sizeOfThenValuesSelector(then.then);
      } else if (then instanceof ChildrenSelector) {
        return this.sizeOfThenChildrenSelector(then.then);
      } else if (then instanceof DescendantsSelector) {
        return this.sizeOfThenDescendantsSelector(then.then);
      } else if (then instanceof FilterSelector) {
        return this.sizeOfThenFilterSelector(then.predicate, then.then);
      }
    }
    throw new WriterException("No Recon serialization for " + then);
  }

  override writeThen(output: Output, then: Value): Writer {
    if (then instanceof Selector) {
      if (then instanceof IdentitySelector) {
        return this.writeThenIdentitySelector(output);
      } else if (then instanceof LiteralSelector) {
        return this.writeThenLiteralSelector(output, then.item, then.then);
      } else if (then instanceof GetSelector) {
        return this.writeThenGetSelector(output, then.item, then.then);
      } else if (then instanceof GetAttrSelector) {
        return this.writeThenGetAttrSelector(output, then.item, then.then);
      } else if (then instanceof GetItemSelector) {
        return this.writeThenGetItemSelector(output, then.item, then.then);
      } else if (then instanceof KeysSelector) {
        return this.writeThenKeysSelector(output, then.then);
      } else if (then instanceof ValuesSelector) {
        return this.writeThenValuesSelector(output, then.then);
      } else if (then instanceof ChildrenSelector) {
        return this.writeThenChildrenSelector(output, then.then);
      } else if (then instanceof DescendantsSelector) {
        return this.writeThenDescendantsSelector(output, then.then);
      } else if (then instanceof FilterSelector) {
        return this.writeThenFilterSelector(output, then.predicate, then.then);
      }
    }
    return Writer.error(new WriterException("No Recon serialization for " + then));
  }
}
